### ISSUE TITLE
Representation and status determination of HOSTSs with multiple NICs (MACs)

### ISSUE THREAD

**Issue by DickenSerm** [jokob-sk/NetAlertX#724]
@DickenSerm hijacking top comment to keep track of this on the top level

> My laptop is recognized as two devices, once via the MAC address of the WLAN card and once via the RJ45 network card. Is it possible to assign two MAC addresses to one device?

__________________

# Proposed Solution and Scope 👨‍💻

> [!IMPORTANT]
>  - Re-use the Devices object as hosts
> - If a user selects a NIC for a Device it will be treated as a host (❓Question: Should only dummy devices be allowed be created as hosts - if so what IP and MAC to use)
> - NICs (Devices with `devHost` not empty) could be automatically hidden from Devices view (tbd maybe replace this with a connection type/entry type DB field drop down) 
> - A new NICs only view under Devices
> - The Network set up and functionality would NOT change
> - ❓ Question: Should we change how notifications are sent out (`down_devices`, `down_reconnected`) and only report on the host once a device is "converted" to a NIC  (= you select a host for it) or should new statuses be created (what would they be)?
> - We need to update how the device card stats are calculated as well or add new options including/excluding hosts
> - TBD if we end up using a specific MAC format, that would allow for Host indentification, we could indicate if Device is a host in menus/listings/the Network layout graph.
> - TBD online status determination of dummy devices - HMW solve for non-host dummy devices? Use case: I want to use ICMP to check online status of an unscannable device - this use case should probably override the default Host status behavior when it has NICs - TBD?

## Device details V1:

![Image](https://github.com/user-attachments/assets/a94e1a69-0e5e-48b6-b9ca-5e4541d88243)

## Device details V2

![Image](https://github.com/user-attachments/assets/c6d9a5ab-b49d-4850-82c8-963acba0041f)

## Other

![Image](https://github.com/user-attachments/assets/9da70021-0983-4061-bba6-eb2d1d3efa70)

![Image](https://github.com/user-attachments/assets/88b7ee7a-621a-4d96-8869-2e7d32606283)

See comment below https://github.com/jokob-sk/NetAlertX/issues/724#issuecomment-2994516655 for details,

**Comment by jokob-sk** [comment]
Hi @DickenSerm,

No, currently that's not possible. MACs are unique identifiers and each new mac generates a new device.

**Comment by DickenSerm** [comment]
I think that's a bit of an unfortunate solution, as the number of devices is rather meaningless.

**Comment by jokob-sk** [comment]
Please suggest alternatives on how to identify a device if you have any
ideas. Most scans return IP or Mac addresses, nothing else. IP addresses
change more frequently than the mac address. I'm open to suggestions.

**Comment by DickenSerm** [comment]
Wouldn't it be possible to say "you are not a device, but a card of this existing device" on a new device? And then have it listed on the existing device as a second MAC address.

**Comment by jokob-sk** [comment]
How would you determine that in an automatic way? Or would that be fully
manual?

**Comment by DickenSerm** [comment]
I was thinking of a manual selection option that would be ignored by default if you don't select anything.

**Comment by jokob-sk** [comment]
So, it would be purely visual, functionality wise, like notifications,
would nothing change?

**Comment by DickenSerm** [comment]
Yes, in principle. My main concern is that the device count is correct.

So I don't have the laptop listed twice under devices and it isn't counted twice.

**Comment by jokob-sk** [comment]
Hummm, I'll think about it. But it's not trivial. There is a lot of
dependent functionality. Cross linking via Mac in the app. The network
tree.

Why is your concern the device count? What count is important and why? Can
you give me one user cases?

I'd rather create a separate view (e.g called Physical devices) than update
other dependencies such as events or the presence views.

**Comment by DickenSerm** [comment]
I find it confusing when 90 devices are shown to me, but in reality I only have 45 physical devices (exaggerated example).

The idea with the extra view for physical devices is a good idea and would be completely sufficient.

But I would then call this view devices and rename the current devices to interfaces or something similar.

**Comment by jokob-sk** [comment]
I'll think about it, but I'm not decided, because I still don't see that
much value in this given the effort.

The more I think about it the less I want to implement it. The UI will
become confusing, as some reports rely on devices, some on interfaces.

Not sure, I'll sleep on it.

If you can give me more use cases, and suggestions how to keep the UI
clean, then that would be great.

**Comment by DickenSerm** [comment]
Glad she even thought about it :)

I don't have any further suggestions right now, but I'll think about it :)

**Comment by vladaurosh** [comment]
I agree with @jokob-sk  
This is network scanner so 2 nics are 2 network devices. 
How do you handle if your RJ45 is disconnected and wlan is connected? How to show in UI which one is offline/online?
Also given that UI has IP address info, how this would look like for a computer with 5 nics?

**Comment by FlyingToto** [comment]
Stepping back a second, the same issue applies to any devices with multiple interfaces like routers...  
IMHO, it would make sense to think of a way to use a compound unique key in the database like:

nodename+mac+ip+hostname (circa old SystemV unix). 
nodename = device name  (1 per machine)...
hostname = name associated to specific IP (1 or more per IP)
mac = 1 per interface

Then in the plugins/screens/reports provide views by nodenames (ie devices) vs by hostnames/ip or MAC... 

This wouldn't be trivial to implement, most likely the core DB needs to be quite updated and most plugins will need some adapters or rewritten... but this should allow to handle any L1-L3 assets...  

of course, one could also ask to consider Layer4... nodename+mac+ip+hostname+PORT 
this would allow to perform further services levels inventories/monitors...

**Comment by sanvalerius** [comment]
Hi, first of all thanks for the work done !

I vote for a feature like this, I have the similar use case, for example a tablet device that present 2 entries (different macs) depending on if it connects to the WiFi 5Ghz or the WiFi 2.4 Ghz..
The @FlyingToto approach seems good to me.

**Comment by stanelie** [comment]
+1 here too, I have several laptops that connect either through wifi, their ethernet port or the ethernet port of their docking station. A way to "group" such devices into 1 unique device would be better to establish "presence".

**Comment by stanelie** [comment]
The way they implement this on a TP-Link Deco router I have is they use this for parental control. You create a profile, say "kid1", and then you assign one or more devices to it. So, it is a distinct list of items. This list would be very useful to display a "person present" list.

**Comment by eagle79** [comment]
I'm not sure this helps with the device count, but I could imagine a Jira issue-styled device "link" element. You could have the ability to add a link with a type from one device to another. One of the link types could be "is same physical device as" or similar. Probably a catch-all link type as well (something like "is related to"). And, of course, the ability for the user to add their own custom link types.

**Comment by jokob-sk** [comment]
@eagle79 thanks, could you elaborate how would this influence the behavior of the listing page(s)?

Also - what other link types would you have and again, how would that influence the app behavior:

>And, of course, the ability for the user to add their own custom link types.

Since you are talking about Jira you can form your answers in this user story format to add more context:

As a <user>  I want to be able <what> so that <reason, use case>

**Comment by eagle79** [comment]
> @eagle79 thanks, could you elaborate how would this influence the behavior of the listing page(s)?
> 
> Also - what other link types would you have and again, how would that influence the app behavior:
> 
> > And, of course, the ability for the user to add their own custom link types.
> 
> Since you are talking about Jira you can form your answers in this user story format to add more context:
> 
> As a I want to be able so that <reason, use case>

To be sure, I don't really need this feature, but the OP seems to want to be able to associate multiple MACs with a single device. Since MAC seems to define uniqueness of a device in the system (which does seem logical to me), that would be difficult. However, I think it's reasonable to want to define links or associations between devices (in the same way you can form links between issues in Jira). So as a use case:

As a user, I want to be able to associate/link a device with one or more other devices, defining what the association/link means. As an example I have a laptop with both a wired and wireless connection, which both have their own MAC. Because this laptop appears as two different devices in the system, I'd like to create a link between them (as a data element on the device details page) that indicates that they are the same physical device (e.g. on device for the wireless MAC,  "This device [is the same as] [John's Laptop (wired)]" (the latter element would be a link to the wired device). Further, I'd like to be able to create such links between any two devices with any customizable meaning I desire (e.g. "This device [is on host] [ESX Server]"). 

In Jira, such links have text for both directions (which _can_ be the same). For example:
* (on BUG-1234) "relates to BUG-1235" and (on BUG-1235) "relates to BUG-1234".
* (on BUG-4321) "duplicates BUG-5321" and (on BUG-5321) "is duplicated by BUG-4321".
See [Linking Issues | Jira Software](https://confluence.atlassian.com/jirasoftwareserver/linking-issues-939938934.html) for more.

I'm thinking the "is the same as" link would be of the former variety (same text in both directions). While this doesn't directly fix the OP's concern about device counts, it at least does allow him to note that a given device is associated with another device.

**Comment by jokob-sk** [comment]
Thanks a lot for the detailed description @eagle79 , appreciated. I'll see what others think about app behavior, because there might be implied expectations. So the more clarity the better. Listing and displaying devices is one thing, influencing notifications is a whole new beast. 

The idea of different links is interesting, but not sure the complexity to build and maintain this forward (maybe not obvious, but it's only me vs Atlassian lol) 

Either way, interesting idea, will keep in mind when deciding on an implementation approach 🤔

**Comment by eagle79** [comment]
I just ran into a slightly different use case where this is problematic, so I thought I'd include it in this discussion. I have a device on my network that apparently generates a new MAC address from time to time, leading NetAlertX to see it as a new device. This is, of course, a slightly different situation than the other use cases, as I really just need a way to merge the two devices under the "new" MAC address. I have seen this with iPhones as well, which seem to do it more regularly (perhaps on each new connection?). 

I'm not sure there's a good solution to this issue (MAC randomization). Certainly, NetAlertX would not be able to detect that they're the same physical device, but having the option as a user to indicate somehow that two different MACs are the same device would be helpful as a way of keeping the device list clean.

**Comment by geurtsd** [comment]
I like this request to - yet in a slightly different (more holistic) view - so +1... 
I miss one aspect (and we may be saying the same thing, yet with different terminology): 
the logical approach misses some steps in the data hierarchy model - and maybe should be able to mirror OSI model more / better:
(PS: disclaimers: 
- I know a little about a lot, yet no expert, so may have said something stupid below - it's meant conceptually and may not be academically correct... ;-)
- I have only recently started using NetAlertX, so maybe that I have not yet truly grasped the setup or capabilities of the tool

1. a MAC is unique, and belongs to 1 NIC/LAN port (RJ45 or wifi) (this is fact and should remain)
2. a NIC can have multiple IP's (most of the time only 1 - though) - or none at all.
3. !!! multiple NICS (in some occasions) can also be "bound" together as 1 IP (although this is more used in pro-environments often by use of LACP) (for load balancing and/or for failover)  - yet would say that this is an exceptional setup - not very often used by users of NetAlertX)
4. An IP can expose several services over multiple ports. (NMAP - plugin sorts this)
5. multiple NIC's can be part of a single HOST (PC, switch, ...etc etc) **(!! the request - the part that I'm missing!!)**
6. a connection between 2 HOSTS goes always over a single link between a single NIC of each of the HOSTS.
(!!most of the times - only exception would be wireless!!)

(exception to point 1 in the wireless space: nowadays, virtual/rotating MAC of wireless devices is a security implementation which completely cuts through this concept of thinking "bottom up" from the OSI layers - this is the issue that @eagle79 is experiencing, I imagine...... FYI: I always disable this feature as I assign static IP's to my mobile devices (yes I know.....) + I have 3 VLAN's implemented to segment traffic out)

So I think the current scanning aspect is good and fairly true (one IP = one MAC = 1 device (I call NIC)) yet I believe it's skipping a step by then define EACH IP-MAC as 1 HOST. I believe a device should be assigned to (child of) a (single) HOST. Where a HOST can have multiple devices (children) (NIC's).....

The (default) scanning of course cannot always detect all of the NICS part of a single host - yet this often is the published devicename (hostname - DNS-name ) (besides via snmp, cdp, lldp, ...) yet this could be built by the owner (like: merge multiple devices (NIC) into 1 host..... then links could be created as is possible now, by linking 1 device to another.

This would also open up the implementation of patch cabinets..... as these are "HOSTS" with a lot of (virtual) devices (each RJ45 port is a device) yet without an IP or MAC..... (so a dummy - non existing MAC could be given (maybe even in a different format - non-compliant to MAC-std - example P1P01 (patch 1 port 01)).......

If this would be possible in the system, It would raise the level of the tool (beyond what it is now) in my view....

again... not an expert in anything, yet know some stuff about a lot (conceptually)... coming from my background in IT.... ;-)
(don't shoot me for errors typed..- its conceptual ;-))

PS: happy to help think this out - or even code a bit - but once again - know a lot about a lot, but no expert ....

**Comment by jcss2503** [comment]
Hi all. The thing is that nowadays all the wifi devices have a feature to ramdomly generate a new mac each time they connect to an AP. This is a very good security feature, without this your device could be tracked each time it **tries** to connect an AP, and that means that your privacy is being compromised. But is clear that this makes the implementation of tools as NetAlerX harder. Other professional products, for example in the OT field use the approach of "device fingerprint". To make the story short that means collect several pieces of information about the device and design an algorithm to generate an unique id (maybe an hash-code?). In this approach a device has multiple pairs of IP,mac assingned. Hope this comment could help to a future development.

**Comment by sdutremble** [comment]
Would SNMP tags be useable to indicate which NIC devices are under which host?

It would require more work setting up SNMP on hosts having multiple NICs but it may provide the logic for this requirement.

**Comment by slammingprogramming** [comment]
Hi @jokob-sk and everyone else,

Thanks to everyone who's added to this thoughtful discussion — I’d like to propose a model that respects NetAlertX’s current architecture while introducing a scalable way to associate multiple MACs with a single logical device or "host."

Primary goal:
Introduce an optional “host profile” abstraction — a user-defined grouping mechanism where multiple MAC-level devices (NICs) can be assigned to a single physical/logical host. The goal is not to disrupt scanning logic but to give users a way to visually and conceptually organize their networks by actual physical devices.

Implementation Approach (Layered by Complexity)

1. Manual Grouping UI

Allow users to manually group devices (i.e., MACs) under a host profile (e.g., “User 1’s Laptop”) via the UI:

Select multiple devices → "Group as Host"

Name the host

Optionally mark one MAC as "primary"

This avoids automatic assumptions while providing control to the user. All existing reports continue to function with MAC-level granularity unless the host profile is explicitly used in some way in the code in future.

A bonus feature this would enable would be that we could show aggregate presence or history at the host level as well as at the NIC level.

2. Device Links (Jira-style)

Allow users to define logical links between devices (as per @eagle79’s suggestion):

“Is same physical device as”

“Connected via switch”

“Is child of host X”

Each link type can be neutral (visual only) or behavioral (affects grouping, counts, etc.). This paves the way for other relationships (VLAN mappings, virtual interfaces, Docker hosts, etc.) without forcing them all into one schema.

3. Optional Passive Fingerprinting

Introduce optional heuristics to suggest groupings via:

Matching hostnames

DNS-resolved names

Consistent IP ranges + OS fingerprints (using nmap, SNMP, etc.)

Repeated MAC–IP switches for the same hostname

Users can approve/reject suggestions, avoiding auto-merges. This system would learn over time and improve merge recommendations without enforcing false positives.

4. Structural Reframing in UI (Optional)

Introduce two views:

Interfaces View (Current): One MAC = One Device

Host View (New): MACs grouped under logical hosts

If desired, users can rename views:

"Devices" → "Interfaces"

"Hosts" → "Devices"
(or leave as-is, depending on comfort level)

You can also color-code/annotate interfaces under the same host.

This solves the following use cases:

Laptops with Ethernet + WiFi + Dock (manual group)

Phones with randomized MACs (manual merge/heuristic merging where possible)

Devices switching NICs (e.g., 2.4GHz vs 5GHz/Wi-Fi vs Ethernet)

Routers with multiple interfaces (visual link)

Reducing device count clutter (host summary view)

This can be done with minimal risk to the current logic as it has no disruption to MAC-level scanning, detection, or alerting, the grouping logic is optional and visual-first, and the reports/UI can default to existing behavior unless grouped view is explicitly used.

This also opens up future improvements and features such as:

Auto-grouping with SNMP tags (as @sdutremble noted)

Hash-based “fingerprints” as a last-resort fallback (as @jcss2503 mentioned)

Host-level presence notifications

Exportable relationship graphs (for patch panels, racks, etc.)

TL;DR

Keep MACs as devices, but allow manual host grouping (with optional smart suggestions). Start simple (manual), allow scaling up to SNMP, fingerprinting, and link graphs later — all without breaking the core detection and reporting logic.

I’d be happy to contribute code, UX mockups, or help vet how much of this could be staged in phases, and will get around to it when I have time.

**Comment by jokob-sk** [comment]
Thanks a lot for the in-depth write-up @slammingprogramming 💪  

The approach I'm currently thinking about is probably a combination of 1 and 2 above. 

> [!NOTE]
> Requirements:
> - Minimal DB structure changes - limiting changes to the Devices DB table
> - Ideally leveraging existing UI functionality (Filters, Settings, Devices listing pages)
> - Not in scope: Presence views updates based on hosts/presence of hosts

## Mockup

![Image](https://github.com/user-attachments/assets/a84384c1-3344-4793-b57e-4ee65630f41d)

### Use cases: 

From my understanding of the user requirements, they fall into 2 main areas:

A. Visual representation and grouping of multiple NICs into one host.
B. Notification related functionality (notify only if the whole host is down).

## Phase 1: Data structure update

1. Introduce an additional field on the Devices table called e.g. `devHost` that would contain the host MAC
1. Allow selection of physical Host device on the above field (filter out current device and any other devices selected as hosts)
1. (optional) Creating a "devNICs" field  that is uneditable but auto updated (determined on the fly?) based on the `devHost` selection on other fields to indicate status of child NICs? [comma separated MAC values? Difficult to work with] - idea by @FlyingToto
1. (optional) Creating a `devRequireAllNicsOnline` Require all NICs online (true = AND, false OR evaluation) -  idea by @FlyingToto

> This will allow the creation of the relationship without extra binding tables. I try to keep everything in the Devices table in order to make backups and data integrity simple.

❓ Question: Should the status of the host be overridden by child NICs? If so how should the host status behave?
❓ Question: Should hosts be always dummy devices and manually created? This would be cleaner than overriding a valid status of a NIC that was marked as a host. TBD: fake MAC format for hosts?
❓ Question: Should NICs (= devices with an non-empty `devHost` field) be by default hidden from the default Devices view? How should these be accessible? Should we create a separate NICS only view?

### 🎯 How might this address use cases: 

A. Visual representation and grouping of multiple NICs into one host. (`somewhat` (visible in the list, we could add a host filter) )
B. Notification related functionality (notify only if the whole host is down). (`n/a`)

## Phase 2: UI changes

> This is the part where I'm hesitating. There are different ways how the above data structure changes can influence the UI. 

Listing from least to most complex:

1. Creating a Hosts filter (similar to Location, Plugin source etc.) [Devices view only]
1. Creating a separate hosts only view, similar to Favorites [Devices view only] - should this include hosts only or hosts + devices without hosts?
1. Visual indication of a host (e.g. icon next to name) [Devices view only]
1. Updating the Network topology view to indicate a host [Network view only] - same icon as above
1. Hiding devices that are selected as children of a host [Devices view only] - needs to be toggleable via  a UI setting

### 🎯 How might this address use cases: 

A. Visual representation and grouping of multiple NICs into one host. (`addressed` - almost fully)
B. Notification related functionality (notify only if the whole host is down). (`n/a`)

## Phase 3: Notification changes

> Work in progress - TBD

To answer:

1. Which notification types should be influenced?
1. Do we need more statuses / notification types?
1. Do we need to thing about specific publishers? E.g. MQTT for home assistant?
1. Do we allow to have per-device setting to toggle between AND/OR NIC status logic indicating online/offline stus of the host? (by @FlyingToto - might be too complex)

### 🎯 How might this address use cases: 

A. Visual representation and grouping of multiple NICs into one host. (`addressed` - almost fully)
B. Notification related functionality (notify only if the whole host is down). (`somewhat`)

# Proposed Solution and Scope 👨‍💻

> [!IMPORTANT]
>  - Re-use the Devices object as hosts
> - If a user selects a NIC for a Device it will be treated as a host (❓Question: Should only dummy devices be allowed be created as hosts - if so what IP and MAC to use)
> - NICs (Devices with `devHost` not empty) could be automatically hidden from Devices view
> - A new NICs only view under Devices
> - The Network set up and functionality would NOT change
> - ❓ Question: Should we change how notifications are sent out (`down_devices`, `down_reconnected`) and only report on the host once a device is "converted" to a NIC  (= you select a host for it) or should new statuses be created (what would they be)?



# Please feedback 📢

I'll keep updating this based on feedback from this thread. Anyone here, please read this comment and add your thoughts and feedback on the above how the suggested approach would address your use cases. 

FYI @eagle79 @stanelie @jcss2503 @sanvalerius @vladaurosh @DickenSerm @slammingprogramming @geurtsd @FlyingToto 

MVP (initial minimum viable product release): Phase 1 + select Phase 2 items (which TBC based on feedback?) - Phase 3 is out of scope for now, might change based on feedback.

**Comment by slammingprogramming** [comment]
I like your ideas, but I'd like to understand *why* some things are out of scope. As the primary creator and maintainer of NetAlertX obviously you have a better understanding of the goals of the software, so I'd like to better understand in order to engage with the content better.

However, I do agree simple backup/restore is essential, and can fully understand how these changes could heavily influence the UI. I argue that that is not necessarily bad, just *tedious*. But I'd argue in the long run a modularized UI is better than hardcoded (idk if it is or not, as I haven't gotten to that part of the codebase yet.) and that the UI honestly could use a little work in some areas, while in others it blows me away. So, I think it would be best to start at Phase 0: design. I'm sure you have a good idea of what you envision, I'd just like to get an idea of what that vision is so I can better align my focuses with the core goals of the project.

**Comment by jokob-sk** [comment]
Hi @slammingprogramming - thanks for the feedback and that's a valid response. 

## Data Backup simplicity

I've found that I was able to help restore way more instances with this simplified backup system (basically 3 files: `app.conf`, `devices.csv`, `workflows.json`) than I would if the database was needed. Some of these are also used e.g. in the SYNC plugin. 

## Modularized UI

I'm very supportive of this. The current UI is OK, but definitely not easy to maintain , looks dated, and feels clunky. However there are probably other things that need to happen first, so here is some context (you will probably discover more things once you dive into the code):

- originally it was a pure php backend but I found it tedious to debug and develop so I switched to JS/python/GraphQL where I could
- the application SQLIte DB is great for backup (that's what I like) but performance is slow so I had to implement some basic caching and move the processing of some tasks to the browser (people try to run this on rapberry Pi 2 )
- I started making all data accessible via the GraphQL server, but things are still missing (e.g. Events) and adding support for more advanced filtering (I think). 
- the caching system and expiration (redis? - currently in browser session). - needs redoing
- It's important to understand how the UI is rendered and that most of it is driven from plugin config.json manifests and settting values (e.g. NEWDEV or CUSTPROD plugins). More details here: https://jokob-sk.github.io/NetAlertX/PLUGINS_DEV/?h=plugin
- I'm at best a mediocre developer so my code will require lots of cleanup and refactoring
- I'd like to use a framework which is on the lighter side as in my experience React-based front ends tend to be very heavy on resources and resulting image sizes are pretty big

For the reasons above, I'd suggest to have a separate discussion on overall UI improvements somewhere else (New Discussion maybe?) as not to delay the implementation of this functionality, with the scope limited to what I mentioned above. Happy to discuss though.

**Comment by slammingprogramming** [comment]
Uhhh.... *I'm* running this on a Raspberry Pi 2 Model B lol

Honestly, I find it makes a great case for the use of efficient methods and SQL calls. I think browser-side processing where appropriate and possible is a great idea for offloading some processing out to another device when viewing the webpages.

Mediocre devs don't produce software of this ability. You're at *least* halfway decent(EDIT FOR CLARITY: I say this in kind, because I actually think your code is just very natural. Spaghetti code is a natural occurrence. I hope my additions/modifications will make it easier to manage for you as well.). I just like standardization so it happens my code often "looks neat" but I have a pile of not-even-yet-started projects myself because of this as well. I'll spend *months* prepping design and such for forward compatibility, scrap it ten times, and then finally really start on it lol.

React will absolutely be a no-go in my opinion. I believe making NetAlertX modular, lightweight, and capable of running on even low-end hardware will be necessary to ensure anyone can enjoy the functions and features, as well as significantly improve expandability and maintainability moving forward.

I don't mind refactoring the codebase for better performance and robustness, and hopefully that will also help you have less weight on your shoulders as I can clearly see the spaghetti code, I do the same thing tbh lol.

I agree, both the UI and also the refactoring/improvement discussions belong separately to both allow development to continue smoothly, but I will be working on that in the background as well to get things running a bit more smoothly within the codebase. 

I'm thinking unit testing implementation would also be a excellent idea to cover common, edge, and invalid cases to ensure robust and reliable code for the backend, and possibly for the frontend down the road.

I think for now your proposed solution would be excellent, and we can build on it down the line to integrate it with a updated UI and such later. I do think it misses some use cases, but for now would significantly improve the ability to handle more use cases that are more common in a more immediate timeframe and allow us to later handle those other cases with less workload on our part. You're definitely thinking iteratively which I think is essential with a program of this size and nature.

**Comment by FlyingToto** [comment]
GM Everyone,  

@jokob-sk ,

> Allow selection of physical Host device on the above field (filter out current device and any other devices selected as hosts)
> This will allow the creation of the relationship without extra binding tables. I try to keep everything in the Devices table in order to make backups and data integrity simple.
> 
> ❓ Question: Should the status of the host be overridden by child NICs? If so how should the host status behave?

that's definitely an interesting question...  may I suggest 3 options (by order of preference)? 
a) if you had the ability to represent each interfaces on separately, maybe you could flag a down interface red and an up interface green while the device itself remains one single color... this could work in the device list view where your status colun would contain several interfaces color-coded accordingly and in the map view...  one advantage is that the logic is already here today since the status is per interface not per machine... 

b) assign one of the interface as "primary" and map the status of the machine to the status of that primary interface. you can most likely identify the "primary" one my mapping the hostname of that interface with the nodename of the machine itself (let the user overwrite it as well)

c) Allow so more granular customizable logic...  like or/and....  allow the user to decide to mark the machine as up if one of interface1 or interface2 is up...  OR, allow him to decide that the host is only up if interface1 AND interface2 is up....  
A typical use case would be: 
a laptop  would be up if my ethernet cable is up OR my wifi inteface is up... 
a dual hosted bridge/firewall is up only if both north/south interfaces are up...  

In all cases, you can introduce the notion of "partially down" where you color orange the ones that have a mix of up or down... 
  


> ❓ Question: Should hosts be always dummy devices and manually created? This would be cleaner than overriding a valid status of a NIC that was marked as a host. TBD: fake MAC format for hosts?

I think there would be some value in clearly segregating the automated portion from the manual one in our dataset....  
for instance:
- color code in white, the data automatically guessed by  NX and its plugins. 
- color code in blue the data manually entered by the user.
- color code in purple the automatically generated data manually-overwritten by the user 
as you introduce the "machine" notion,  it is likely to be mostly manual on day 1, then as we go along automate more and more....

**Comment by FlyingToto** [comment]
one small trick which might help with the graph view: 
(I haven't coded a GUI in 30 years and it was all in tckltk so no idea of the feasibility)
Since it is very difficult to represent an actual network with loops (especially with backup links between switches). 

if you had the ability to create frames...  you could have a base network view of your routers/switches/access-points which doesn't show the devices per say only the networks.  and when someone click on one of the network open a new frame showing the devices on this lan only. any machine dual hosted can show a second link going out to a different LAN icon...  
etc...

**Comment by jokob-sk** [comment]
Hi @FlyingToto , thanks for your input. This is a quick mockup (note making the Host NICs editable would be too difficult):

![Image](https://github.com/user-attachments/assets/74d7965d-2c5a-4674-a126-fe010e2004df)

For clarification, each interface will be a separate device - as it i snow, what this would add is another relationship between them. Similar to the existing Network node field, however NICs would not be shown in the Network diagram (as devices currently are once you add them). I pop-up might be possible later on. 

To clarify this question:

❓ Question: Should hosts be always dummy devices and manually created? This would be cleaner than overriding a valid status of a NIC that was marked as a host. TBD: fake MAC format for hosts?

- this would mean you could not convert a device to a Host - all hosts would have to be manually created with the Create new device functionality.

**Comment by FlyingToto** [comment]
Oh I see, so the "host" is more like a group of 1 or more NICs... the problem is that it is going to be a pain to enter them manually... 

what if you basically create host per NIC automatically, but then give the user to group several NICs under the one of the group? 
for instance: 

1) auto-generate:
nic1=mylaptopname.fqdn.lan=192.168.0.10=01:02:03:04:05:06:07:08 mapped to host1=mylaptopname.fqdn.lan
nic2=mylaptopnamewifi.fqdn.lan=192.168.0.11=01:02:03:04:05:06:07:09 mapped to host2=mylaptopnamewifi.fqdn.lan 
nic3=somethingelse.fqdn.lan mapped to host3=somethingelse.fqdn.lan


2) allow the user to map nic2 to host1 manually:
nic1=mylaptopname.fqdn.lan=192.168.0.10=01:02:03:04:05:06:07:08 mapped to host1=mylaptopname.fqdn.lan
nic2=mylaptopnamewifi.fqdn.lan=192.168.0.11=01:02:03:04:05:06:07:09 mapped to host1=mylaptopname.fqdn.lan
so now.  host1 is mapped to 2 devices with their own MAC, hostnames, etc...

**Comment by jokob-sk** [comment]
Thanks fir the input @FlyingToto 

I don't think that every device should create a separate host - this will initially double the number of devices and over complicate things in my opinion. I'd argue that that there are not that many needs for host devices in most networks. Usually 1 NIC = 1 device. We might add auto generation of hosts later, but it would be initially out of scope to keep things easy to roll out, test and iterate on. If we'd go with generating an extra host this would double devices (I know one user has ~3000 devices - this would kill the performance instantly). We'd have to rewrite all listing pages (Events, Presence, Devices, Possibly the Network topology page) and that would blow out scope.

**Comment by FlyingToto** [comment]
really, the way I see it, this host column would just be one extra field and it would be mostly static so does it really has much of an impact?

**Comment by jokob-sk** [comment]
Sorry unsure what you are referring to. Each host would be represented as an entry in the Devices database table.

**Comment by FlyingToto** [comment]
oh I was under the impression that you would basically rename the "device" table into a "NIC" table with one extra column which would be the host(machine) the nic is attached to...   that way you can keep 99% of the NAX as-is on day 1.  
then on day 2, you would give 2 versions of the graph, one by nic (like we have today), one by host/machine  maybe with the LANs consolidated with an option to open them in a separate frame etc... 

but do you plan to have the device list contain rows which could be either a single NIC or a host with multiple NICs? that is going to be much more complicated isn't it?

**Comment by jokob-sk** [comment]
Thanks for the follow up. See the requirements in the above comment where only the Devices table is edited:

https://github.com/jokob-sk/NetAlertX/issues/724#issuecomment-2994516655

I'm not planning to create additional DB tables due to the above reasoning. 

Also, in the first implementation no changes to the graph (I assume that you reference the network topology graph) are planned, again see the above comment for details. 

> but do you plan to have the device list contain rows which could be either a single NIC or a host with multiple NICs?

Yes, see proposed solution scope above:

> NICs (Devices with devHost not empty) could be automatically hidden from Devices view
> A new NICs only view under Devices

> that is going to be much more complicated isn't it?

In contrary, this would require minimal DB and UI changes as I'd leverage existing UI and DB structures, which are integrated with other functionality, such as backups, CSV import / export, multi edit and the SYNC plugin.

Hope this explanation makes sense.

**Comment by FlyingToto** [comment]
hehehe, a picture is worth a thousand words... :-) can you show us what it would look like in the device list?  using my very basic example above what you are suggesting is basically going to look like this in the device view: 

![Image](https://github.com/user-attachments/assets/50737940-336e-42ad-af72-8dbc23c6042e)


What I was thinking was more in line with something like this:
![Image](https://github.com/user-attachments/assets/39716d06-2a54-4f30-89e3-0ad8b2f2645b)

![Image](https://github.com/user-attachments/assets/04b6f6d7-a848-4f66-8728-87877131a4a6)
the main advantage being that it seems a bit more intuitive in my mind to list all the machines 1 per row for most people and use an actual nodename as a key instead of the mac address... (if we can't figure the nodename programmatically allow people to manually set it) internally you still uise the MAC as the primary key in your table but for the human reading it, they would use the nodemae... kind of a human readable alias associated with your devhost? 


of course, to add insult to injury, I am pretty sure that one can assign the same nodename to more than one machine... but then again, I can also float IP addresses amongst different machines, SUN servers are notorious to have the same MAC address for all their NIC by default etc..., etc... :-)

**Comment by jokob-sk** [comment]
It could look e.g. this way in the Devices list: 

![Image](https://github.com/user-attachments/assets/95d36631-13c2-4efa-a676-9dab2f9f37f5)

NIC<1-4> = Current Device name of a device representing a NIC

**Comment by FlyingToto** [comment]
oh ok, so  you use devhost as the key (using the MAC) but the your present the data by names like in my second table...  and you still give use another view by NIC equivalent to today's list correct?

**Comment by FlyingToto** [comment]
oh one more thing, I just thought of something... how are switches and access points be represented? 

- if you have a 48 ports switch you would have 48 NICs... 
- if you have an access point, you have dynamic number of clients if you show the devices connected to it? 
--> I answered my own question, it seems that an AccessPoint will have one MAC per SSID they call them BSSID...

**Comment by jokob-sk** [comment]
Just to reiterate - MAC uniqueness is a core design principle of the app which would be incredibly hard to change. 

Unsure if I understand this:

> .. and you still give use another view by NIC equivalent to today's list correct?

![Image](https://github.com/user-attachments/assets/9bcb25aa-92a0-4e37-893d-3f332caee048)

It's mentioned here, to add another view for NICs only:

# Proposed Solution and Scope 👨‍💻

> [!IMPORTANT]
>  - Re-use the Devices object as hosts
> - If a user selects a NIC for a Device it will be treated as a host (❓Question: Should only dummy devices be allowed be created as hosts - if so what IP and MAC to use)
> - NICs (Devices with `devHost` not empty) could be automatically hidden from Devices view
> - A new NICs only view under Devices 👈
> - The Network set up and functionality would NOT change 
> - ❓ Question: Should we change how notifications are sent out (`down_devices`, `down_reconnected`) and only report on the host once a device is "converted" to a NIC  (= you select a host for it) or should new statuses be created (what would they be)?
> - We need to update how the device card stats are calculated as well or add new options including/excluding hosts


This would not change as it is done now with the network construction - you already can represent  switches and access points in the current set of features. The NICs assignment would be represented as mentioned above.

> oh one more thing, I just thought of something... how are switches and access points be represented?
>
>    if you have a 48 ports switch you would have 48 NICs...
>    if you have an access point, you have dynamic number of clients if you show the devices connected to it?

**Comment by geurtsd** [comment]
Hey Guys,

great to see this moving forward... or at least triggering briljant thoughts.... :-)

My reply may be a bit messy, as its reaction to the chain over the last 2 days.... yet bear with me....

I like @slammingprogramming proposal of the "host profile" abstraction approach.... this would give the flexibility to group NIC's into a logical grouping..... (be it mobile, physical or a combination thereof)......
it would provide a lolistic answer to all potential devices out there that "connect" in one way or another (Ethernet or otherwise even - think of BT, 4G, 5G, LORA, .....)

Most communication technologies kinda resonate against the first layers of the OSI model, and the proposed way is then holistic enough, also for the future.... 

I would lean more to the "manual" logic mapping instead of "out of the box" smart-stuff..... (at first).... potentially plug-ins could be developed going forward to enable automation based on use-cases..... yet would propose "less is more" in this case to start with and gather experiences in the field..... 

As to @jokob-sk write-up on his envisioned approach:
- the use-cases for me apply both, so both visual as well as (avoiding) false downtime notification 

> 1. Introduce an additional field on the Devices table called e.g. devHost that would contain the host MAC

I'm failing to understand the reasoning to introduce a MAC on host-level as its an attribute of a NIC, and hence would be inherited by the host (in a 1-multiple relationship) - OR do you just want to list the "primary MAC" here - so not as a attribute, but as a relational field, linking to the primary key field of an existing NIC (so selectable - not "free fill")

as for the phasing: whatever is feasible, I can agree to.... the phasing sounds logical.... and any progress is good progress, in the (voluntary) open-source world.. ;-)

To the data-structure questions:
> ❓ Question: Should the status of the host be overridden by child NICs? If so how should the host status behave?

if you mean status as in "presence": then yes... and I would say for logic: if any of the NIC's is online, the Host is online.  
> ❓ Question: Should hosts be always dummy devices and manually created? This would be cleaner than overriding a valid status of a NIC that was marked as a host. TBD: fake MAC format for hosts?

I would say, yes, dummy for now. Automation can come in later-on via plug-ins (and use-case based), but they would then still be "manual via a plug-in"..... the "Fake MAC" for a host would not be needed in my view (logically).... 
unless the MAC field in your current setup is NEEDED/mandatory (technically), and then I would propose to use a single fake MAC from the [assigned private ranges,](https://www.blackmanticore.com/fc5c95c7c2e29e262ec89c539852f8fb#:~:text=Black%20Manticore&text=Like%20private%20IP%20address%20ranges,xx%2Dxx%2Dxx%2Dxx) as a workaround, yet the more clean solution would be not to make it mandatory for a host (as its not an attribute of a host, yet of a "child-NIC".)

>❓ Question: Should NICs (= devices with an non-empty devHost field) be by default hidden from the default Devices view? How should these be accessible? Should we create a separate NICS only view?

Optionally: yes, and easiest would be indeed creating 2 views: one with empty field, and one with filled field..... However I see this going into the direction of 2 user groups:
- those that will adopt the hosts, and they will have no independent NICS, as this would be orphaned NICS (and thus indicate "errors" in their data Quality)
- those that will not adopt HOSTS, and they will not have any hosts, as they have no need for it.

**Comment by jokob-sk** [comment]
Thanks @geurtsd for the detailed write up - I will digest it fully tomorrow morning (midnight here). However, an important thing to realize regarding MACs - maybe you guys can think of something while I sleep 🙂

> I'm failing to understand the reasoning to introduce a MAC on host-level as its an attribute of a NIC, and hence would be inherited by the host (in a 1-multiple relationship) - OR do you just want to list the "primary MAC" here - so not as a attribute, but as a relational field, linking to the primary key field of an existing NIC (so selectable - not "free fill")

- MAC uniqueness is a core design principle of the app which would be incredibly hard to change. So a Device representing host needs to have a unique MAC (as any other NICs). So dummy Hosts need to make a fake or a reserved mac address format (as with `Internet`). It's not pretty but I work with what I inherited.

**Comment by geurtsd** [comment]
@jokob-sk ... enjoy your R&R... 😉 

as far as I see it:
The "simplest" solution to remain within the design principles would be to adopt a mandatory "auto-assigning" of MAC addresses to any created (dummy) host. This means writing a function that assigns an (unused) MAC out of a certain range.

as for the range(s) I see 2 options out of below defined (standard) ranges: 

![Image](https://github.com/user-attachments/assets/df8f9619-f72d-445d-8be9-680e197a6919)

From Above righthand column (as you do not ever want to have a "collision" with a MAC that is used "out there" on a device): 
1. using the unicast MAC ranges:
These are reserved and unique (should be) on any LAN.... they are not allowed to be used by any manufacturer of HW (to identify such hardware).
However, some (more and more) providers have adopted this also as the function to "randomize" MAC addresses from a security (obscuring) perspective. - so you potentially could still get duplicates that way as the assigned MAC in NetAlertX is not "seen" by the randomisation function of the device (as it does not reply to ARP / NDP requests)....  

2. using the Multicast MAC ranges:
Not a beautiful solution, but you would never get a "Collision" in overlap, as these ranges are never used by a nic (fixed or random) as they cover multiple MAC's in the relaying of packest (so never a unique NIC)....

I think in practice, using the multicast ranges would be the simplest solution (for now)....

All other solutions would either entail redesigning your DB structure (additional host table) or redesigning your function and stepping away from the requirement of "MUST have MAC" and/or "MUST be UNIQUE". These solutions entail more work, time and coding, while (mis)using the Multicast MAC range is just a "design agreement" within NetAlertX. It would not be perfect (=non-compliant with standard), yet workable.... .

if needed, you could even hide the MAC of the HOST and only use it (under the hood) to comply with your design-principles.....

**Comment by FlyingToto** [comment]
@geurtsd ,
Fundamentally, NAX was designed to identify MAC addresses on a given LAN. So the data-structure uses the MAC as a primary key... that's why when we link PC **foo** to port 10 of switch **bar**, we actually use the MAC address of **bar** and the port number, not the name...  

Now what @jokob-sk is proposing is a fairly elegant way to grant us the ability to present actual machines with more than one NIC without going crazy by switching to completely new structure indexed on nodenames...  it reminds me of the way in LDAP you can either define a group **guux** as a list of users **u, uu, uux** or add a memberof=guux  attribute to each individual user records of u, uu, and uux individually.... in our case our dataset is relatively small so this does the job very well.. 

in this case, we normalement don't even need to use dummy MAC since most devices we care about already have one. However, it is a great option when we want people to be able to create assets (or "NIC") where the MAC is hidden... for instance my cablemodem only shows me its south interface, I have no way to find the MAC adddress on the coax side... 

note about the local adminisstered MAC: these are the ones used for randomized MAC so we kind of need to differentiate the ones we generate as placeholders vs the ones coming from real devices.. 
using mutlicast address might not work when nax is used in a lab or something 
one simple alternative since for the most part we would use this pseudo MAC as key only is simply to use something that is not MAC compliant... like a GX-XX-XX-XX-XX-XX-XX-XX since G is not a valid Hex...

**Comment by jokob-sk** [comment]
@geurtsd @FlyingToto - thanks a lot for your inputs and enlightenment regarding locally administered and reserved MAC ranges. 

I like the designation of a locally administered multicast MAC to represent a Host. Thinking of e.g. having the  `XF-XX-XX-XX-XX-XX` range reserved for Hosts.

❓ Question: Doesn't the Network Node (`devParentMAC`) already mimic the functionality of a `devHost` field? For those unfamiliar check this [Network guide](https://jokob-sk.github.io/NetAlertX/NETWORK_TREE/?h=netw).

❓ Question: If we need both `devParentMAC` and a new `devHost` can someone explain to me why?

If that would be the case then what's missing is the filtering, display of NICs (children nodes) and some notification update work, to put it simply. 

Also BTW for those who don't know, you can already create dummy devices ([see docs](https://jokob-sk.github.io/NetAlertX/DEVICE_MANAGEMENT/?h=dummy#dummy-devices)) with a random MAC function (this would be changed to adhere to the `XF-XX-XX-XX-XX-XX`  format.

**Comment by FlyingToto** [comment]
❓ Question: Doesn't the Network Node (devParentMAC) already mimic the functionality of a devHost field? For those unfamiliar check this 
devParentMAC links a device  foo to its Parent switch that it is connected to isn't it?  
devHost links MAC abc to MAC def which is part of the same host so on your graph they should show as the same device not separate ones linked by a  line

❓ Question: If we need both devParentMAC and a new devHost can someone explain to me why?
as per above...   
1) device A would be connected by a line to device B tracked by its devParentMAC
2) device A  would show as a single entity on your graph as a secondary NIC under the existing device mapped to irs devHost

**Comment by jokob-sk** [comment]
Thanks @FlyingToto 

> devHost links MAC abc to MAC def which is part of the same host so on your graph they should show as the same device not separate ones linked by a line

To re-iterrate above: 

> Also, in the first implementation no changes to the graph (I assume that you reference the network topology graph) are planned, again see the above comment for details.

The library used to generate the graph doesn't have other capabilities than what's available. E.g. only one line from a bubble to another. 

Also we are talking only about the data structure here, there is already currently only a MAC-to-MAC link available, and this won't change with this new devHost field - lets forget the visual representation for now (e.g. I could already combine the existing Network topology data into such a representation). Happy for you to mock something up though.

**Comment by jokob-sk** [comment]
Hummm thinking more about my above comment on if we need a devHost field, maybe instead we need a drop down field on devXyzType (not sure what xyz is yet, maybe host, or entry) to select if this entry (device) is a host, nic, Stand alone device. This would then influence how the entry would be handled. Then we could reuse the devParentMac field and just change what the connection signifies (a link between devices, a link between a host or nics).

Not sure if my thinking here is correct though.

**Comment by geurtsd** [comment]
hey @jokob-sk, @FlyingToto,

Thanks for the background.... Fully understand the design starting point and principle now.... will try to answer and "think along" within those concepts going forward.. 😉 

As for the remarks about the MAC: I agree with @FlyingToto that a fake MAC  is even a better solution (did not think it was possible - Data validation rules or such), as that would never exist.
just start with GG-XX-XX-XX-XX-XX... that's even more beautiful than mis-using valid MAC's... this would make it clear for anybody that its a fake MAC...... (with some technical background)..... 

I concur with the conclusion of @FlyingToto that the devParentMac serves a different purpose as being intended by devHost..... 

As for your last proposal of splitting / identifying a host via a seperate device: that would be possible... and you could re-use the field in that case.... However its not as beautifull... and (very theoreticaly speaking): I would advise against it from a DB normalisation standpoint..... 
Next to the theoretical (cleanness) aspect: repurposing (using) a field and giving it 2 purposes / meanings will complicate any code that is based on this field (be it functional or display related)..... as you would always have to check the type to understand what the field means and filter out the ones that you do not want..... 

but again, its do-able.... every choice has its pro's and con's.....

**Comment by jokob-sk** [comment]
I definitely don't want to use one field for two purposes, but it would simplify the display of relationships (already accounted for in the Network graph for example).

I'm asking if you need two separate relationships between NICs and a host. Isn't the the Network node relationship then obsolete? If so, it would be more difficult to create validation for e. G. Only allowing one of the relationships to be specified on a Nic or host or device.

❓Question : what is a use case where you have a different  network parent node than a nic host parent relationship. Hope the question makes sense. 

> repurposing (using) a field and giving it 2 purposes / meanings will complicate any code that is based on this field

That's a fair concern. I do not see however great difficulties in using this approach in code. Thinking on the fly: if link type selected as NIC, hide from default devices view,if device or host to host, show.

This then could easily also be used to dynamically generate a list of children NICs or Hosts/devices on the parent device without duplicating the UI logic. 

Having an existing (one field) for the relationship is also easier to set up and you don't have to double up the work, which is better UX in my opinion when setting things up. 

Happy to be educated of course.

**Comment by geurtsd** [comment]
Understand that re-using the field is the "easiest" solution.... as said, every choice has pros and cons..... 

>I'm asking if you need two separate relationships between NICs and a host. Isn't the the Network node relationship then obsolete? If so, it would be more difficult to create validation for e. G. Only allowing one of the relationships to be specified on a Nic or host or device.

I do not see a required relationship between multiple NIC's and a host (at least in the sense of a relationship as in the design, being a "line" between NIC and host). A line between NIC-NIC is OK.
its more a grouping...... Just to visualize:
this is "one" host in my setup (yes a complicated one as its unraid with dockers):

![Image](https://github.com/user-attachments/assets/4fd3545a-9364-4dca-8e25-ae1213b5a472)

However it would be nice if this would be grouped somehow in the overview to indicate they "belong together".

>❓Question : what is a use case where you have a different network parent node than a nic host parent relationship. Hope the question makes sense.

This question triggers me, and actually lets me think that you cannot re-purpose the "devParentMac".... as it would be set at the level of the NIC, not the HOST..... yet in this setup, it has to be able to contain multiple values.

(in wording: a "NIC-C" can be connected to another parent "NIC-P", AND that same "NIC-C" can be part of "HOST-C".....)
So thats 2 MAC's required in the same field.... + then you cannot distinct anymore which is for "linking", and which is for "hosting"

if you want to build the HOST in between 2 nics in the hierarchy, this would break your automatic inventory plugins, I imagine.

Only option would be to use the "devParentMac" of a dummy host, and list in there multiple "child-NICS". 
This way it remains "independent" from all the other logic..... and has least impact....

"brainfart": building on your suggestion to create a custom type .... if you define a type "host".... does the tooling have an option to define fields specific (and only) for that type?
Then you could create a "ContainsNICs" field and list the devices in there which are part of that host.

for the visualisation: what is the library used for the graph?

**Comment by jokob-sk** [comment]
Thanks for the reply @geurtsd and patience explaining these things 🙏 

Your use case where you (I assume) want to hide all these entries under one host would be supported based on my proposal:

> this is "one" host in my setup (yes a complicated one as its unraid with dockers): ...

Steps to configure this:

✅ - indicates what is already possible
🆕 - needs to be implemented

1. (optional if you don't have one) Create dummy Host by clicking New device ✅ (might have future enforcement of the GG- MAC)
1. Navigate to each NAS.local entries and set `devParentMac` (= Network node) to the above (you can use multi edit) ✅
1. Set the `devRelationshipType` (TBD)  to e.g. NIC (we could add virtual machine, host etc) (you will be able to use multi edit) 🆕 
1. Devices with certain type of relationship woudl be automatically hidden from the default view (there could be even a setting `UI_hide_relationship_types` where the user can select which will be hidden from the default view) 🆕
1. If a host device has the relationship type NIC, it would then apply conditional status determination for it 🆕


What I'm not clear on (could you please elaborate):

> This question triggers me, and actually lets me think that you cannot re-purpose the "devParentMac".... as it would be set at the level of the NIC, not the HOST..... yet in this setup, it has to be able to contain multiple values.

Maybe a draw.io diagram would help me to understand your concern.

> "NIC-C" can be connected to another parent "NIC-P", AND that same "NIC-C" can be part of "HOST-C".....

Aren't only NIC to NIC connections possible? Wouldn't this be represented the following way?
- HOST-C
  - NIC-P devParentMAC: HOST-C, devRelationshipType: `NIC`
    - NIC-C devParentMAC: NIC-P, devRelationshipType: `NIC`

The proposed approach would also support this:

- HOST-C
  - NIC-P devParentMAC: HOST-C, devRelationshipType: `NIC`
    - HOST-B devParentMAC: NIC-P, devRelationshipType: `simple` (TBD)
      - NIC-C devParentMAC: HOST-B, devRelationshipType: `NIC`

This way you can also inter lay hosts and group multiple NICs and hosts under the same parent. 

- HOST-C
  - NIC-P devParentMAC: HOST-C, devRelationshipType: `NIC`
  - HOST-B devParentMAC: HOST-C, devRelationshipType: `simple` (TBD)
    - NIC-C devParentMAC: HOST-B, devRelationshipType: `NIC`

> "brainfart": building on your suggestion to create a custom type .... if you define a type "host".... does the tooling have an option to define fields specific (and only) for that type?

No, but I could hide it conditionally in the UI.

> for the visualisation: what is the library used for the graph?

treeviz.js: https://github.com/PierreCapo/treeviz (free to use for FOSS, usable with vanilla JS, didn't find suitable alternatives yet with more functionality)

Sorry for brevity, getting ready for work

**Comment by FlyingToto** [comment]
from my side, we can probably adapt to either options. I do think that overall, it would be easier for everyone if  NaX datastructure was a bit more self-explanatory. so in my mind, we call a device an actual physical thing  (a router, a switch, a firewall, a pc, a server,...) with 1 or more MAC.... now we still auto-discover and auto-build mostly from: 
1) the MAC addresses 
2) the DNS names 
3) the extra-fancy plugins
....
4) the manual intervention...

**Comment by geurtsd** [comment]
@jokob-sk ... thx, sorry for the longer wait in reply... sometimes busy... 😄 

>Your use case where you (I assume) want to hide all these entries under one host would be supported based on my proposal:

The hiding is not needed for me..... 

>Maybe a draw.io diagram would help me to understand your concern.

my remark is not as much about "preference" or "not working", yet more out of concern that it may break (other) functionality as the setup is not alligned to networking standards.....

below some drawings as you suggested:

This is how it works today, and also how (any) network scanning devices would "see" the setup in its tables / after scans:
![Image](https://github.com/user-attachments/assets/3ad94118-1de9-4ad7-a28b-da75e47c6ccc)

If you would implement the use of a dummy host "VIA" repurposing the "devParentMac", it would show like this:
![Image](https://github.com/user-attachments/assets/447bc631-285e-4aa2-ba2e-cff3854492a5)
!!!! this would potentially break plug-ins as they cannot scan for the dummy HOST record as parent.... Additionally, scanning plugins may automatically update any records (on example Device 1 back to "ROUTER MAC") as in the results of the network scan it sees a connection to the router and not the HOST..... 

If you would implement the "devHost" field and only use it on the dummy hosts, it would look like this:
![Image](https://github.com/user-attachments/assets/542d4abf-4afe-4bf0-a95a-85d3e3fcca94)
!!! this would NOT break any automatic Scanning

PS: a better name would then also be "devChilds" iso "devHost", imho.

**Comment by jokob-sk** [comment]
Thanks a lot for the detailed explanation @geurtsd and no need to apologize for the wait

> If you would implement the use of a dummy host "VIA" repurposing the "devParentMac", it would show like this:

FYI the setup already supports dummy devices: https://jokob-sk.github.io/NetAlertX/DEVICE_MANAGEMENT/?h=dummy#dummy-devices

> !!!! this would potentially break plug-ins as they cannot scan for the dummy HOST record as parent.

Scanning of dummy devices is already supported - e.g. if you add a loopback IP address, such as 0.0.0.0, it will always show as online. With the proposed changes, the status of such a device could be determined by the statuses of the nics associated with a dummy device. 

> PS: a better name would then also be "devChilds" iso "devHost", imho.

Its really hard to handle fields with multiple values (e.g. in an array or comma-separated) and create relationships with it within SQL queries so I try to keep 1 value per field, that's why I'm leaning to using the devParentMAC ( or potentially devHost) field. You can still represent a parent-child relationship, just the information about the relationship is saved on the child. 

>  Additionally, scanning plugins may automatically update any records (on example Device 1 back to "ROUTER MAC") as in the results of the network scan it sees a connection to the router and not the HOST.....

This might happen, not sure how often though - currently only the UNIFI and OMADA plugins auto-import the topology setup. Let's say I'd go with the relationship-type field approach (devRelationshipType) - the plugins can default building the hierarchy with a `default` relationship type and if the user wants, he can aggregate devices where it makes sense into a `host-nic` relationship.

> it would show like this:...

Just to re-iterate, the initial scope is showing the relationships on a device, not in the network diagram, which will remain unchanged (maybe I'll manage to show nics as icons in the device bubbles, but not sure yet).

With the approach of relationship types, here are the configuration options to a user (they can choose if they want to configure their setup with or without fake devices ( - reminder - fake devices are already possible, they just don't support NICs and notification changes).

![Image](https://github.com/user-attachments/assets/c2e2c798-71cf-4ea7-ae4b-b21ec55203d7) 

(You can edit the diagram by downloading it from here: https://pastebin.com/ZZ8jCR5W and then using https://app.diagrams.net to edit/annotate)

**Comment by geurtsd** [comment]
hey @jokob-sk,

maybe some clarification required: 
When I say "scanning" I mean scanning (taking inventory) out of the real-world devices registers. which is then used as input to "fill" NAX.
So the dummy host cannot be scanned "out of" the real world devices, as they (the real-world device registers) do not know about it..... 

UNIFI is indeed one of those, which I use (and can test with if needed).

SO my remark is against this scanning, that the re-use of the field devParentMac would break this functionality most likely.... (and maybe more, its the only I can think of at the moment, without having the full NAX logic insightful) .... 
PS: I just have a huge dislike for breaking regression in coding (wanted or unwanted) from my real world experience in IT.... it always creates more troubles than it solves (in my experience).... 

as to the 2 drawio's created... I prefer the one to (only) supporte the dummy host creation for the use case.... its more logical and closer to "human readable".... although the second one works also, its not as intuitive and straightforward, again potentially complicating (or breaking) regression, as well as future features.

**Comment by jokob-sk** [comment]
Hi @geurtsd , thanks a lot for the read and comments!

> When I say "scanning" I mean scanning (taking inventory) out of the real-world devices registers. which is then used as input to "fill" NAX.
> So the dummy host cannot be scanned "out of" the real world devices, as they (the real-world device registers) do not know about it.....

That is true, dummy devices can't be scanned. However you can fake the availability of them (presence) with a loopback IP + ICMP plugin setup, or as per plans outlined above, could be based on the NICs presence (based on user's config and decisions). 

> SO my remark is against this scanning, that the re-use of the field devParentMac would break this functionality most likely.... (and maybe more, its the only I can think of at the moment, without having the full NAX logic insightful) ....
> PS: I just have a huge dislike for breaking regression in coding (wanted or unwanted) from my real world experience in IT.... it always creates more troubles than it solves (in my experience)....


Thanks for raising the concern. I carefully considered the reuse of `devParentMac` field in the current proposal. No auto-assignment is planned for the first release. In the future it may be for new detections under specific matching conditions (e.g., same hostname/vendor/IP range).

Existing logic that uses `devParentMac` — like topology — won't be affected, as I ensure not to overwrite user-defined values or use it in conflicting ways. 

That said, I fully agree about avoiding regressions. I always do my best to make sure there is an upgrade path and no change in existing installation behavior. If you can point to any specific functionality you're worried might be impacted, I’d be happy to dig into it further.

Making sure I consider these concerns is why I started this discussion in this issue in the first place, to test my ideas against a group of people to see if there are any issues in my logic 😄That's why I really appreciate the time and thought you put into responding.

**Comment by geurtsd** [comment]
@jokob-sk ... let me know which route you ultimately decide to go.... if you need testing or validation, happy to help in my environment.... - also since I have Unifi / Ubiquity  implemented for scanning

**Comment by jokob-sk** [comment]
A first iteration of this is available in the dev image.

It would be great if you could test this. Can you please switch to the `netalertx-dev` docker image (backup everything at first or, ideally, spin up a **separate container with a separate db and config** as DB modifications can happen), in about 15 minutes (or after the latest [action finishes](https://github.com/jokob-sk/NetAlertX/actions)) from now?

Make sure you refresh your browser cache - and click the 🔄 refresh button in the top right corner.

<img width="1764" height="1146" alt="Image" src="https://github.com/user-attachments/assets/2749c6e9-fcca-476a-8e7c-b77c7cd70280" />

<img width="363" height="459" alt="Image" src="https://github.com/user-attachments/assets/8361b0bc-071c-454f-9711-fde0d3f5505f" />

<img width="1314" height="93" alt="Image" src="https://github.com/user-attachments/assets/c19b6627-0bb3-4654-a8d9-eb000c78cfd7" />

Please note the UI and functionality may still change. 

Thanks in advance,
j

**Comment by jokob-sk** [comment]
Refined it a bit further: 

<img width="1634" height="1271" alt="Image" src="https://github.com/user-attachments/assets/8a47a468-3a36-4071-95c4-edde8bd80a2a" />

EDIT: Added colors

<img width="476" height="90" alt="Image" src="https://github.com/user-attachments/assets/b1768d16-ed68-40e4-9e32-8a276a30baf1" />

<img width="868" height="813" alt="Image" src="https://github.com/user-attachments/assets/87286b58-e4f5-47b7-9981-d3d63b23841f" />

**Comment by geurtsd** [comment]
@jokob-sk,

will test... have it installed allready... may take a bit of time - busy times)....

**Comment by jokob-sk** [comment]
Hi @eagle79 @stanelie @jcss2503 @sanvalerius @vladaurosh @DickenSerm @slammingprogramming @geurtsd @FlyingToto - if you can feedback on the implementation in the `netalertx-dev` image in the next few days so I can stop working on this if it's ready for release and focus on a different initiative. 

Thanks in advance,
j

**Comment by eagle79** [comment]
> If you can feedback on the implementation in the `netalertx-dev` image in the next few days so I can stop working on this if it's ready for release and focus on a different initiative.

I'm sorry, I'm not going to be able to look at it for a bit... Covered up with other things. I'll take a look when I can, but please don't wait for my feedback. The solution sounds good as described.

**Comment by jokob-sk** [comment]
releasing in 15 min -> closing this issue
